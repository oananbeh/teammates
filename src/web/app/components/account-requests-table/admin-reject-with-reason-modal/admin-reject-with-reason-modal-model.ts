export interface RejectWithReasonModalComponentResult {
    rejectionReasonTitle: string;
    rejectionReasonBody: string;
}
