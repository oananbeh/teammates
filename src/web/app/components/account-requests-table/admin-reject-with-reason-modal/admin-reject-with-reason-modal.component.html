<div class="modal-header bg-warning">
    <h5 class="modal-title">
      <div>Reject Account Request for {{ accountRequestName }} With Reason</div>
    </h5>
    <button type="button" class="btn-close" (click)="activeModal.dismiss()"></button>
</div>
<div id="reject-account-request-modal" class="modal-body">
  <div class="row">
    <div class="col-12">
      <div>
        <label><b>Rejection Message Title:</b></label>
      </div>
      <div class="form-group">
        <input id="rejection-reason-title" type="text" class="form-control" rows="5" value="{{ rejectionReasonTitle }}" (input)="rejectionReasonTitle = $event.target.value">
      </div>
      <div>
        <label><b>Rejection Message Body:</b></label>
      </div>
      <tm-rich-text-editor id="rejection-reason-body" [richText]="rejectionReasonBody" (richTextChange)="onRejectionReasonBodyChange($event)"></tm-rich-text-editor>
</div>
<div class="modal-footer">
  <button id="btn-cancel-reject-request" type="button" class="btn btn-light" (click)="activeModal.dismiss()">Cancel</button>
  <button id="btn-confirm-reject-request" type="button" class="btn btn-warning" (click)="reject()">
      Reject
  </button>
</div>
