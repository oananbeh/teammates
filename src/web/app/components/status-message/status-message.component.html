<div class="alert alert-{{ statusMessage.color }}" *ngFor="let statusMessage of messages">
  <button class="btn-close" (click)="dismiss(statusMessage)"></button>
  <div style="padding-right: 30px;" [innerHtml]="statusMessage.message"></div>
</div>
