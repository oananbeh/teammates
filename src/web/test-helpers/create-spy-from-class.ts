/**
 * Creates a spy object for a given class, providing spy functions for all of its methods.
 *
 * @param Class - The class for which to create the spy
 * @returns A spy object with spy functions for each method of the provided class.
 * @example
 * class MyClass {
 *   method1() {}
 *   method2() {}
 * }
 *
 * createSpyFromClass(MyClass) returns { method1: jest.fn(), method2: jest.fn() }
 */
export default function createSpyFromClass<T>(Class: new (...args: any[]) => T): Spy<T> {
  const spy = {} as Spy<T>;

  Object.getOwnPropertyNames(Class.prototype).forEach((methodName) => {
    if (typeof Class.prototype[methodName] === 'function') {
      spy[methodName as keyof T] = jest.fn();
    }
  });

  return spy;
}

// Helper type for better typing
type Spy<T> = {
  [K in keyof T]: jest.Mock<any, any>;
};
