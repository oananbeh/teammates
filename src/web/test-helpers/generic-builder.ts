import { Course, Instructor, JoinState, Student } from '../types/api-output';

type GenericBuilder<T> = {
  [K in keyof T]: (value: T[K]) => GenericBuilder<T>;
} & { build(): T };

/**
 * A generic builder function that creates a builder object for constructing objects with specified initial values.
 *
 * @template T - The type of object being constructed.
 * @param initialValues - The initial values for the object being constructed.
 * @returns A generic builder object.
 * @example
 * // Create a course builder with initial values.
 * const courseBuilder = createBuilder<Course>({
 *   courseId: 'exampleId',
 *   courseName: '',
 *   timeZone: '',
 *   institute: '',
 *   creationTimestamp: 0,
 *   deletionTimestamp: 0,
 * });
 *
 * // Usage of builder:
 * const myCourse = courseBuilder
 *   .courseName('Introduction to TypeScript')
 *   .timeZone('UTC+0')
 *   .institute('My University')
 *   .creationTimestamp(Date.now())
 *   .build();
 */
export function createBuilder<T extends object>(initialValues: T): GenericBuilder<T> {
  const builder: any = {};

  (Object.keys(initialValues) as (keyof T)[]).forEach((key) => {
    builder[key] = (value: T[keyof T]) => {
      initialValues[key] = value;
      return builder;
    };
  });

  builder.build = () => ({ ...initialValues });

  return builder;
}

export const courseBuilder = createBuilder<Course>({
  courseId: 'exampleId',
  courseName: '',
  timeZone: '',
  institute: '',
  creationTimestamp: 0,
  deletionTimestamp: 0,
});

export const instructorBuilder = createBuilder<Instructor>({
  courseId: 'exampleId',
  email: '',
  name: '',
  joinState: JoinState.JOINED,
});

export const studentBuilder = createBuilder<Student>({
  courseId: 'exampleId',
  email: 'examplestudent@gmail.com',
  name: 'test-student',
  teamName: 'test-team-name',
  sectionName: 'test-section-name',
});
