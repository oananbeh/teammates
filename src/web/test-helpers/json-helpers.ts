/**
 * Store helper functions for using JSON.
 */

export const mapReplacer = (_key: any, value: any): any => {
  if (value instanceof Map) {
    return {
      dataType: 'Map',
      value: Array.from(value.entries()),
    };
  }
  return value;
};

export const mapReviver = (_key: any, value: any): any => {
  if (typeof value === 'object' && value !== null) {
    if (value.dataType === 'Map') {
      return new Map(value.value);
    }
  }
  return value;
};
