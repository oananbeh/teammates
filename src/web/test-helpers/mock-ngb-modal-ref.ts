import { NgbModalRef } from '@ng-bootstrap/ng-bootstrap';

/**
 * Mock NgbModalRef by implementing only the most important fields.
 */
export const createMockNgbModalRef =
    (componentInstance: any = {}, result: Promise<any> = Promise.resolve()): NgbModalRef => {
      return {
        result,
        get componentInstance(): any {
          return componentInstance;
        },
      } as NgbModalRef;
    };
