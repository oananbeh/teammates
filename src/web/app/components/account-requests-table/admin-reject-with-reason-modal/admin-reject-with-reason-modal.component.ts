import { Component, Input, OnInit } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { RejectWithReasonModalComponentResult } from './admin-reject-with-reason-modal-model';
import { environment } from '../../../../environments/environment';
import { AdminSearchResult, InstructorAccountSearchResult, SearchService } from '../../../../services/search.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { ErrorMessageOutput } from '../../../error-message-output';

/**
 * Modal to select reject account requests with reason.
 */
@Component({
  selector: 'tm-reject-with-reason-modal',
  templateUrl: './admin-reject-with-reason-modal.component.html',
  styleUrls: ['./admin-reject-with-reason-modal.component.scss'],
})

export class RejectWithReasonModalComponent implements OnInit {

  @Input()
  accountRequestName: string = '';

  @Input()
  accountRequestEmail: string = '';

  existingAccount: InstructorAccountSearchResult = {
    name: '',
    email: '',
    googleId: '',
    courseId: '',
    courseName: '',
    isCourseDeleted: false,
    institute: '',
    courseJoinLink: '',
    homePageLink: '',
    manageAccountLink: '',
    showLinks: false,
    awaitingSessions: {},
    openSessions: {},
    notOpenSessions: {},
    publishedSessions: {},
  };

  rejectionReasonBody: string = '<p>Hi, {accountRequestName} </p>\n\n'
  + '<p>Thanks for your interest in using TEAMMATES. '
  + 'We are unable to create a TEAMMATES instructor account for you.</p>'
  + '<p><strong>Reason:</strong> The email address you provided is not an &#39;official&#39; '
  + 'email address provided by your institution.<br />'
  + '<strong>Remedy:</strong> Please re-submit an account request with your &#39;official&#39; '
  + 'institution email address.</p>\n\n'
  + '<p><strong>Reason:</strong> The email address you have provided seems like it belongs to a student '
  + '(i.e., not a staff member) of your institution.<br />'
  + '<strong>Remedy:</strong> If you are a student but you still need an instructor account, '
  + 'please send your justification to {supportEmail}</p>\n\n'
  + '<p><strong>Reason:</strong> You already have an account for this email address and this institution.<br />'
  + '<strong>Remedy:</strong> You can login to TEAMMATES using your Google account: {googleId} </p>\n\n'
  + '<p>If you are logged into multiple Google accounts, remember to logout from other Google accounts first, '
  + 'or use an incognito Browser window. Let us know (with a screenshot) if that doesn\'t work.</p>'
  + '<p>If you need further clarification or would like to appeal this decision, please '
  + 'feel free to contact us at {supportEmail}</p>'
  + '<p>Regards,<br />TEAMMATES Team.</p>';
  rejectionReasonTitle: string = 'We are Unable to Create an Account for you';

  constructor(
    public activeModal: NgbActiveModal,
    public statusMessageService: StatusMessageService,

    private searchService: SearchService,
  ) {}

  ngOnInit(): void {
    this.rejectionReasonBody = this.rejectionReasonBody.replace('{accountRequestName}', this.accountRequestName);
    this.rejectionReasonBody = this.rejectionReasonBody.replaceAll('{supportEmail}', environment.supportEmail);

    this.replaceGoogleId();
  }

  onRejectionReasonBodyChange(updatedText: string): void {
    this.rejectionReasonBody = updatedText;
  }

  replaceGoogleId(): void {
    this.searchService.searchAdmin(this.accountRequestEmail)
    .subscribe({
      next: (resp: AdminSearchResult) => {
        const hasInstructors: boolean = !!(resp.instructors && resp.instructors.length);

        if (!hasInstructors) {
          this.rejectionReasonBody = this.rejectionReasonBody.replace('{googleId}', 'NO_GOOGLEID');
          return;
        }

        for (const instructor of resp.instructors) {
          if (instructor.googleId !== '') {
            this.existingAccount = instructor;
          }
        }

        if (this.existingAccount.googleId === '') {
          // When an instructor account exists, but for some reason does not have a googleId
          this.rejectionReasonBody = this.rejectionReasonBody.replace('{googleId}', 'NO_GOOGLEID');
        } else {
          this.rejectionReasonBody = this.rejectionReasonBody.replace('{googleId}', this.existingAccount.googleId);
        }
      },
      error: (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    });
  }

  /**
   * Fires the reject event.
   */
  reject(): void {

    if (!this.rejectionReasonBody || this.rejectionReasonBody.length === 0) {
      this.statusMessageService.showErrorToast('Please provide an email body for the rejection email.');
      return;
    }

    if (!this.rejectionReasonTitle || this.rejectionReasonTitle.length === 0) {
      this.statusMessageService.showErrorToast('Please provide a title for the rejection email.');
      return;
    }

      const result: RejectWithReasonModalComponentResult = {
        rejectionReasonTitle: this.rejectionReasonTitle,
        rejectionReasonBody: this.rejectionReasonBody,
      };

      this.activeModal.close(result);
  }
}
