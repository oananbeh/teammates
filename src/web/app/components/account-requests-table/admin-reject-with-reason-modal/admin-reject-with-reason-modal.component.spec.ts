import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';
import { RejectWithReasonModalComponent } from './admin-reject-with-reason-modal.component';
import {
  FeedbackSessionsGroup, InstructorAccountSearchResult,
  SearchService,
} from '../../../../services/search.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { createBuilder } from '../../../../test-helpers/generic-builder';

const DEFAULT_FEEDBACK_SESSION_GROUP: FeedbackSessionsGroup = {
  sessionName: {
    feedbackSessionUrl: 'sessionUrl',
    startTime: 'startTime',
    endTime: 'endTime',
  },
};

const instructorAccountSearchResultBuilder = createBuilder<InstructorAccountSearchResult>({
  name: 'name',
  email: 'email',
  googleId: 'googleId',
  courseId: 'courseId',
  courseName: 'courseName',
  isCourseDeleted: false,
  institute: 'institute',
  courseJoinLink: 'courseJoinLink',
  homePageLink: 'homePageLink',
  manageAccountLink: 'manageAccountLink',
  showLinks: false,
  awaitingSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
  openSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
  notOpenSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
  publishedSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
});

describe('RejectWithReasonModal', () => {
  let searchService: SearchService;
  let statusMessageService: StatusMessageService;
  let fixture: ComponentFixture<RejectWithReasonModalComponent>;
  let component: RejectWithReasonModalComponent;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [],
      imports: [
        HttpClientTestingModule,
        RouterTestingModule,
      ],
      providers: [NgbActiveModal, SearchService, StatusMessageService],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(RejectWithReasonModalComponent);
    searchService = TestBed.inject(SearchService);
    statusMessageService = TestBed.inject(StatusMessageService);
    fixture.detectChanges();
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should show empty title and body', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('replaceGoogleId: should set the googleId to an empty string if no instructor accounts are found', () => {
    jest.spyOn(searchService, 'searchAdmin').mockReturnValue(of({
      students: [],
      instructors: [],
      accountRequests: [],
    }));

    component.replaceGoogleId();

    expect(component.existingAccount.googleId).toEqual('');
  });

  it('replaceGoogleId: should set the googleId to the instructor accounts googleId '
  + 'if an instructor account is found', () => {
    const testInstructor = instructorAccountSearchResultBuilder.googleId('instructorGoogleId').build();

    jest.spyOn(searchService, 'searchAdmin').mockReturnValue(of({
      students: [],
      instructors: [testInstructor],
      accountRequests: [],
    }));

    component.replaceGoogleId();

    expect(component.existingAccount.googleId).toEqual('instructorGoogleId');
  });

  it('reject: should show error message when title is empty upon submitting', () => {
    component.rejectionReasonTitle = '';
    fixture.detectChanges();

    const spyStatusMessageService = jest.spyOn(statusMessageService, 'showErrorToast')
        .mockImplementation((args: string) => {
          expect(args).toEqual('Please provide a title for the rejection email.');
    });

    const rejectButton: any = fixture.debugElement.query(By.css('#btn-confirm-reject-request'));
    rejectButton.nativeElement.click();

    expect(spyStatusMessageService).toHaveBeenCalled();
  });

  it('reject: should show error message when body is empty upon submitting', () => {
    component.rejectionReasonBody = '';
    fixture.detectChanges();

    const spyStatusMessageService = jest.spyOn(statusMessageService, 'showErrorToast')
        .mockImplementation((args: string) => {
          expect(args).toEqual('Please provide an email body for the rejection email.');
    });

    const rejectButton: any = fixture.debugElement.query(By.css('#btn-confirm-reject-request'));
    rejectButton.nativeElement.click();
    expect(spyStatusMessageService).toHaveBeenCalled();
  });

  it('reject: should close modal with data', () => {
    const spyActiveModal = jest.spyOn(component.activeModal, 'close');
    component.rejectionReasonTitle = 'Rejection Title';
    component.rejectionReasonBody = 'Rejection Body';
    fixture.detectChanges();
    component.reject();
    expect(spyActiveModal).toHaveBeenCalled();
    expect(spyActiveModal).toHaveBeenCalledWith({
      rejectionReasonTitle: 'Rejection Title',
      rejectionReasonBody: 'Rejection Body',
    });
  });
});
