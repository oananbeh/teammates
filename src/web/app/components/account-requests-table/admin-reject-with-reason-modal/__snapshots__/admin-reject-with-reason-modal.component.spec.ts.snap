// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RejectWithReasonModal should show empty title and body 1`] = `
<tm-reject-with-reason-modal
  accountRequestEmail=""
  accountRequestName=""
  activeModal={[Function NgbActiveModal]}
  existingAccount={[Function Object]}
  rejectionReasonBody={[Function String]}
  rejectionReasonTitle={[Function String]}
  searchService={[Function SearchService]}
  statusMessageService={[Function StatusMessageService]}
>
  <div
    class="modal-header bg-warning"
  >
    <h5
      class="modal-title"
    >
      <div>
        Reject Account Request for  With Reason
      </div>
    </h5>
    <button
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="reject-account-request-modal"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div>
          <label>
            <b>
              Rejection Message Title:
            </b>
          </label>
        </div>
        <div
          class="form-group"
        >
          <input
            class="form-control"
            id="rejection-reason-title"
            rows="5"
            type="text"
          />
        </div>
        <div>
          <label>
            <b>
              Rejection Message Body:
            </b>
          </label>
        </div>
        <tm-rich-text-editor
          id="rejection-reason-body"
        />
      </div>
      <div
        class="modal-footer"
      >
        <button
          class="btn btn-light"
          id="btn-cancel-reject-request"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-warning"
          id="btn-confirm-reject-request"
          type="button"
        >
           Reject 
        </button>
      </div>
    </div>
  </div>
</tm-reject-with-reason-modal>
`;
